[OPA Plugin] Support splitting OPA filter requests in batch mode by batch size

thinaih opened the pull request:
<!-- Thank you for submitting a pull request! Find more information 
at https://trino.io/development/process.html, 
at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md 
and contact us on #core-dev in Slack. -->
<!-- Provide an overview for maintainers and reviewers. -->
## Description

This PR aims to address issue #25748

<!-- Provide details that help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can.
More info at https://trino.io/development/process#release-note -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( X ) Release notes are required. Please propose a release note for me.
() Release notes are required, with the following suggested text:

```markdown
## Section
* Fix some things. ({issue}`issuenumber`)
```

Praveen2112 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaConfig.java:
```diff
@@ -61,8 +66,25 @@ public void testExplicitPropertyMappings()
+                .put("opa.policy.uri", "https://opa.example.com")
+                .put("opa.policy.batch-size", "0")
+                .buildOrThrow()).build(OpaConfig.class))
+                .isInstanceOf(ConfigurationException.class);
```
Can we match the error message as well ?

Praveen2112 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/RequestTestUtilities.java:
```diff
@@ -34,17 +34,20 @@ private RequestTestUtilities() {}
-                        throw new AssertionError("Cannot parse expected request", e);
-                    }
-                })
+                .map(RequestTestUtilities::toJsonNode)
```
nit: Can we move method `toJsonNode` below `assertStringRequestsEqual` ?

Praveen2112 commented on plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaHttpClient.java:
```diff
@@ -158,25 +162,30 @@ public <T> Set<T> batchFilterFromOpa(Collection<T> items, Function<List<T>, OpaQ
+                (entry, result) ->
+                        Optional.of(requireNonNullElse(result.result(), ImmutableList.<Integer>of()))
+                                .flatMap(indices -> indices.isEmpty() ? Optional.empty() : Optional.of(indices))
+                                .map(indices -> indices.stream().map(index -> entry.getValue().get(index)).collect(toImmutableSet()))
```
nit: Can we move `.map` to a new line ? So as to reduce the diff

Praveen2112 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
Can we have a prep commit for initial changes on these tests ? It would be easier to review on how things would evolve ?

thinaih replied:
Split into two commits one for test refactoring only, another one for the implementation

Praveen2112 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
```diff
@@ -49,226 +52,236 @@
+                        """.formatted(identity.getUser()));
+
+        List<TestCase> testCases = ImmutableList.of(
+                new TestCase(OPA_CONFIG_NO_CHUNKING, ImmutableMap.of(11, 1),1),
```
nit: whitespace after `, 1);`

Praveen2112 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
```diff
@@ -49,226 +52,236 @@
+        );
+
+        for (TestCase testCase : testCases) {
+            assertAccessControlMethodBehaviour(
+                    (accessControl, systemSecurityContext, identities) -> accessControl.filterViewQueryOwnedBy(systemSecurityContext.getIdentity(), identities),
+                    inputIdentities,
+                    identityToJsonNode,
+                    new ExpectedRequestProperties(
+                            testCase.batchSizeToBatchNumber,
+                            testCase.numberOfRequest,
+                            "FilterViewQueryOwnedBy",
+                            inputIdentities.stream().map(identityToJsonNode).collect(toImmutableSet())),
+                    testCase.opaConfig);
```
Can we avoid using this abstraction and rather have these values explicitly like 

```
assertAccessControlMethodBehaviour(situation1);
assertAccessControlMethodBehaviour(situation2);
...
```

So it is easier to understand on what are the impacts

thinaih replied:
Do you mean clarifying if you mean the TestCase abstraction or ExpectedRequestProperties? I assume you mean TestCase

mosiac1 replied:
From what I've seen in previous PRs, what Praveen is refering to is doing something like

```java
@Test
public void testFilterViewQueryOwnedBy()
{
    ...
    assertFilterViewQueryOwnedBy(new TestCase(OPA_CONFIG_NO_BATCH_SIZE, ImmutableMap.of(11, 1), 1));
    ...
}

void assertFilterViewQueryOwnedBy(TestCase testCase)
{
            assertAccessControlMethodBehaviour(
                    (accessControl, systemSecurityContext, identities) -> accessControl.filterViewQueryOwnedBy(systemSecurityContext.getIdentity(), identities),
                    inputIdentities,
                    identityToJsonNode,
                    new ExpectedRequestProperties(
                            testCase.batchSizeToBatchNumber,
                            testCase.numberOfRequest,
                            "FilterViewQueryOwnedBy",
                            inputIdentities.stream().map(identityToJsonNode).collect(toImmutableSet())),
                    testCase.opaConfig);
}
```

The nice thing about this is that when a case fails the traceback will point you the line that failed, so you don't have to figure out which element in the for loop failed

Praveen2112 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
```diff
@@ -278,88 +291,145 @@ public void testFilterColumnErrorCases()
                 "Failed to query OPA backend");
     }
 
+    private static void assertRequestOperationMatches(ImmutableList<JsonNode> requests, String expectedOperation) {
```
nit: `{` to next line. `List<JsonNode>`

mosiac1 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
```diff
@@ -49,226 +52,235 @@
-                .put(tableOne, ImmutableSet.of())
-                .put(tableTwo, ImmutableSet.of())
-                .buildOrThrow();
+        Map<SchemaTableName, Set<String>> requestedColumns = ImmutableMap.of(
+                tableOne, ImmutableSet.of()
+                , tableTwo, ImmutableSet.of());
```
Formatting here is a bit odd
```suggestion
        Map<SchemaTableName, Set<String>> requestedColumns = ImmutableMap.of(
                tableOne, ImmutableSet.of(),
                tableTwo, ImmutableSet.of());
```

mosiac1 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
```diff
@@ -49,226 +52,235 @@
+            InstrumentedHttpClient mockClient = createMockHttpClient(
+                    OPA_SERVER_BATCH_URI,
+                    buildHandler(pathToResources,
+                            ImmutableSet.of("\"table_one_column_1\"", "\"table_one_column_2\"", "\"table_one_column_3\"", "\"table_two_column_1\"")
+                                    .stream()
+                                    .map(RequestTestUtilities::toJsonNode)
+                                    .collect(toImmutableSet())));
```
nit: we could add a `toJsonNode` util that takes an object so we don't have to double quote the string

mosiac1 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
```diff
@@ -278,88 +290,151 @@ public void testFilterColumnErrorCases()
+        });
+    }
+
+    record TestCase(OpaConfig opaConfig, Map<Integer, Integer> batchSizeToBatchNumber, Integer numberOfRequest) {}
+
+    record ExpectedRequestProperties(Map<Integer, Integer> batchChunkSizes, int numberOfRequests, String operation, Set<JsonNode> expectedRequestItems) {}
```
nit: check `requireNonNull(...)` and copy `Map`s and `Set`s in constructor

mosiac1 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
```diff
@@ -278,88 +290,151 @@ public void testFilterColumnErrorCases()
+                objectsToFilter.stream().filter(obj -> objectsToFilter.indexOf(obj) % 2 == 0).collect(toImmutableSet()),
+                objectsToFilter.stream().filter(obj -> objectsToFilter.indexOf(obj) % 2 != 0).collect(toImmutableSet()));
+
+        record SubTestCase<T>(Set<T> expectedResourcesResult, Set<JsonNode> allowedResources)
+        {
+
+        }
```
nit: check `requireNonNull(...)` and copy `Map`s and `Set`s in constructor

mosiac1 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
```diff
@@ -278,88 +290,151 @@ public void testFilterColumnErrorCases()
                         TEST_SECURITY_CONTEXT,
                         "my_catalog",
-                        ImmutableMap.of(new SchemaTableName("some_schema", "some_table"), ImmutableSet.of("column_one", "column_two"))),
+                        ImmutableMap.of(new SchemaTableName("some_schema", "some_table"),
+                                ImmutableSet.of("column_one", "column_two"))),
```
i wouldn't change the formatting here

thinaih committed:
Refactor tests to prepare for the implementation of batch size for opa batch mode

Format

thinaih committed:
Implement batch size for opa bacth mode

thinaih committed:
CR:formatting

CR: perform validation for records, and copy sets and maps

CR: remove test cases loop and replace with explicit method calls

refactor to improve tests readability

chenjian2664 commented on plugin/trino-opa/src/test/java/io/trino/plugin/opa/TestOpaBatchAccessControlFiltering.java:
```diff
@@ -286,80 +424,182 @@ void testFilterColumnErrorCases()
+                ImmutableSet.of(objectsToFilter.getFirst()),
+                objectsToFilter.stream().collect(toImmutableSet()),
+                objectsToFilter.stream().filter(obj -> objectsToFilter.indexOf(obj) % 2 == 0).collect(toImmutableSet()),
+                objectsToFilter.stream().filter(obj -> objectsToFilter.indexOf(obj) % 2 != 0).collect(toImmutableSet()));
```
Could you please add comment what/why is doing here? it's not obvious to me
